torch/csrc/jit/passes/specialize_autogradzero.cpp

```diff
@@ -146,22 +148,35 @@ struct AutogradZeroSpecializer {
       }
 
       state_[inp] = *pttp->undefined() ? State::Zero : State::Nonzero;
-      auto check = graph_->insert(prim::AutogradAnyNonZero, {inp});
+
       if (*pttp->undefined()) {
-        check = graph_->insert(aten::__not__, {check});
+        zero_values.push_back(inp);
+      } else {
+        nonzero_values.push_back(inp);
       }
-      checks.push_back(check);
     }
 
     // unable to specialize any of the inputs
-    if (checks.size() == 0) {
+    if (nonzero_values.size() == 0 && zero_values.size() == 0) {
       GRAPH_DUMP("Unable to add any specialization guards", graph_);
       versioning_if->destroy();
       // the checks we inserted will be cleaned up
       // by any subsequent DCE pass
       return nullptr;
     }
 
+    Value* nonzero_values_list =
+        graph_
+            ->insertNode(graph_->createList(TensorType::get(), nonzero_values))
+            ->output();
+    checks.push_back(
+        graph_->insert(prim::AutogradAnyNonZero, {nonzero_values_list}));
```

Comment:
Do we need to construct a list if we're calling a variadic op?

Reply:
A good question, I'll check.